crates/matrix-sdk/src/encryption/mod.rs

```diff
@@ -622,7 +623,26 @@ impl Client {
                 self.keys_query(r.request_id(), request.device_keys.clone()).await?;
             }
             AnyOutgoingRequest::KeysUpload(request) => {
-                self.keys_upload(r.request_id(), request).await?;
+                self.keys_upload(r.request_id(), request).await.inspect_err(|e| {
+                    match e.as_ruma_api_error() {
+                        Some(RumaApiError::ClientApi(e)) if e.status_code == 400 => {
+                            if let ErrorBody::Standard { message, .. } = &e.body {
+                                // This is one of the nastiest errors we can have. The server
+                                // telling us that we already have a one-time key uploaded means
+                                // that we forgot about some of our one-time keys. This will lead to
+                                // UTDs.
+                                if message.starts_with("One time key") {
```

Comment:
Is this by any chance synapse specific? Would it maybe be better to check for the request's `one_time_keys` list instead? 🤔

Reply:
Just meant to only check requests that contain one time keys in their payload, we don't about errors on any of the others, but as long as we check the string it doesn't really matter, does it.

Reply:
Sorry, still not sure what you mean here. :sweat_smile:

I'll try to answer anyways. 

We are only logging this for `/keys/upload` requests, the first `/keys/upload` request might not contain any `one-time` keys but every subsequent request will.

So it's not particularly useful to check if the request contained one-time keys, the first request a client ever does by definition can't fail with this error either.

Reply:
> but every subsequent request will.

Oh, that was the bit of info I was missing. I was under the impression one time keys don't get used and replenished that often.

Reply:
That depends on how many new devices want to talk to you, but this requests purpose is to replenish and upload new one-time keys to the server.